### PR TITLE
Adding reference counting to Cached Connections

### PR DESCRIPTION
**Pull request by j-clap**

### CODE REVIEW

File: bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/util/ReferenceCountedHashMapTest.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import org.junit.Test;
+
+/** Created by j-clap on 10/17/19. */
+public class ReferenceCountedHashMapTest {
+
+  private final String[] KEYS = {"first", "second"};
+  private final Integer[] VALUES = {1, 2};
+
+  @Test
+  public void test_insertion_works() {
```
Comment (rahulKQL) [inline bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/util/ReferenceCountedHashMapTest.java]: We follow the camelcase naming convention for test case names for this repository. Could you please update these names accordingly.

File: bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/util/ReferenceCountedHashMapTest.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    Map<String, Integer> testMap = new ReferenceCountedHashMap<>();
+    testMap.put(KEYS[0], VALUES[0]);
+    Assert.assertEquals(testMap.get(KEYS[0]), VALUES[0]);
```
Comment (rahulKQL) [inline bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/util/ReferenceCountedHashMapTest.java]: It'd be nice to statically import `assertEquals` & `assertNull`

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java [hunk 3]
```diff
@@ -733,6 +735,10 @@ public synchronized void close() throws IOException {
       throw new IOException("Could not close the admin client", ex);
     }
 
+    if (options.useCachedChannel() && options.useGCJClient()) {
+      cachedClientContexts.remove(options.getDataHost());
```
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java]: As discussed offline, this isn't enough. The background resources in the ClientContext need to be shutdown

Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java]: As discussed offline, this isn't enough. The background resources in the ClientContext need to be shutdown

Comment (j-clap) [reply bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java]: Added new changes to address :)

File: bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/util/ReferenceCountedHashMapTest.java (hunk 2 above)
Comment (j-clap) [reply bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/util/ReferenceCountedHashMapTest.java]: I'm just following the same conventions as the other tests in Utils.

File: bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/util/ReferenceCountedHashMapTest.java (hunk 1 above)
Comment (j-clap) [reply bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/util/ReferenceCountedHashMapTest.java]: I'm just following the same conventions as the other tests in Utils.

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/util/ReferenceCountedHashMap.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+ * there is more functionality which could be used from Java8, we have chosen to not subclass them
+ * as we are not expecting their use as of now. Other items have not been subclassed as they do not
+ * affect reference counts (e.g.: get).
+ */
+@InternalApi
+public class ReferenceCountedHashMap<K, V> extends HashMap<K, V> {
+
+  // As removal operates via object reference, we must keep this map as Object as
+  //  opposed to K typed.
+  private Map<Object, Integer> counterMap;
```
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/util/ReferenceCountedHashMap.java]: please mark as final

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/util/ReferenceCountedHashMap.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+  private Callable<V> cleanerCallback;
```
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/util/ReferenceCountedHashMap.java]: final

Comment (j-clap) [reply bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/util/ReferenceCountedHashMap.java]: done

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/util/ReferenceCountedHashMap.java (hunk 4 above)
Comment (j-clap) [reply bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/util/ReferenceCountedHashMap.java]: done

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/util/ReferenceCountedHashMap.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+   */
+  public synchronized V remove(Object key) {
+    // Normally I would throw an exception on not found, but hash map doesn't
+    //  so I'll follow suit.
+    if (!super.containsKey(key)) {
+      return null;
+    }
+    Integer currentCount = this.updateReference(key, -1);
+    if (currentCount == 0) {
+      counterMap.remove(key);
```
[the remark this answers is not in the document]
Comment (j-clap) [reply bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/util/ReferenceCountedHashMap.java]: done

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/util/ReferenceCountedHashMap.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+  /**
+   * Remove removes a reference to the key within the context of the reference count. If no
+   * references are left, it also removes the pair from the underlying HashMap.
+   *
+   * @param key The key for which the reference should be removed.
+   * @return The value removed, or null if it does not exist.
+   */
+  public synchronized V remove(Object key) {
+    // Normally I would throw an exception on not found, but hash map doesn't
+    //  so I'll follow suit.
```
[the remark this answers is not in the document]
Comment (j-clap) [reply bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/util/ReferenceCountedHashMap.java]: done

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/util/ReferenceCountedHashMap.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+  /**
+   * PutAll puts several values into the HashMap adding references for each. Note that since this is
+   * iterating through the map twice, it is not the most efficient algorithm. The implementation of
+   * this method is kept for compatibility purposes, but its use is not recommended.
+   *
+   * @param m A map of all the values to be added.
+   */
+  public void putAll(Map<? extends K, ? extends V> m) {
+    // Map iterations are inherently slow. As such the use of this
```
[the remark this answers is not in the document]
Comment (j-clap) [reply bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/util/ReferenceCountedHashMap.java]: done

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/util/ReferenceCountedHashMap.java [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+   *
+   * @param key The key reference to be updated.
+   * @param updateCount The value to update it to.
+   * @return The final count of the reference post update.
+   */
+  private Integer updateReference(Object key, Integer updateCount) {
+    // Not making this synchronized as the methods calling this are expected to be synchronized.
+    //  No reason to double lock (note: compiler might eliminate, but unsure)
+    Integer currentCounter = counterMap.get(key);
+    currentCounter = updateCount + ((currentCounter == null) ? 0 : currentCounter);
```
[the remark this answers is not in the document]
Comment (j-clap) [reply bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/util/ReferenceCountedHashMap.java]: done

**Pull request merged**